pkg/typesystem/weighted_graph.go

```diff
@@ -0,0 +1,138 @@
+// its siblings edges for intersection based via the weighted graph.
+// If the direct edges have equal weight as its sibling edges, it will choose
+// the direct edges as preference.
+// If any of the children are not connected, it will return empty IntersectionEdges.
+func GetEdgesForIntersection(edges []*graph.WeightedAuthorizationModelEdge, sourceType string) (IntersectionEdges, error) {
+	if len(edges) < 2 {
+		// Intersection by definition must have at least 2 children
+		return IntersectionEdges{}, fmt.Errorf("invalid edges for source type %s", sourceType)
+	}
+
+	directEdges := make([]*graph.WeightedAuthorizationModelEdge, 0, len(edges))
+	var lowestEdge *graph.WeightedAuthorizationModelEdge
+	directEdgesAreLowest := false
+	lowestWeight := 0
+
+	// Process all edges first to separate direct from non-direct
+	nonDirectEdges := make([]*graph.WeightedAuthorizationModelEdge, 0, len(edges))
+
+	// It is assumed that direct edge always appear first.
+	for _, edge := range edges {
+		if !hasPathTo(edge, sourceType) {
+			continue
+		}
+		if edge.GetEdgeType() == graph.DirectEdge {
+			// The first step is to establish the largest weight for all the direct edges as they need to be
+			// treated as a group. This weight will serve as the baseline for other siblings to compare against.
+			directEdges = append(directEdges, edge)
+			// hasPathTo already verified the weight exists, so we can safely ignore ok
+			if weight, _ := edge.GetWeight(sourceType); weight > lowestWeight {
+				directEdgesAreLowest = true
+				lowestWeight = weight
+			}
+		} else {
+			nonDirectEdges = append(nonDirectEdges, edge)
+		}
+	}
+
+	if lowestWeight == 0 {
+		if edges[0].GetEdgeType() == graph.DirectEdge {
+			// this means that all the direct edges are not connected.
+			// In reality, should not happen because the caller should have trimmed
+			// the parent node already.
+			return IntersectionEdges{}, nil
+		}
+	}
+	for edgeNum, edge := range nonDirectEdges {
+		if weight, _ := edge.GetWeight(sourceType); weight < lowestWeight || (edgeNum == 0 && len(directEdges) == 0) {
+			// in the case of the first edge, we need to initialize as the least weight
+			// for other edges to compare against it.
+			// For non-first edges, only replace if it has the lowest weight.
+			lowestEdge = edge
+			lowestWeight = weight
+			directEdgesAreLowest = false
+		}
+	}
+
+	siblings := make([]*graph.WeightedAuthorizationModelEdge, 0, len(edges)-len(directEdges))
+
+	// Now, assign all the non directly assigned edges that are not the lowest
+	// weight to the sibling edges. Even if the directly assigned edges are not the lowest
+	// weight, we want to treat these directly assigned edges separately. The reason is that
+	// if one of these directly assigned edge satisfy, the list objects candidate may still
+	// be valid.
+	for _, edge := range edges {
+		if edge.GetEdgeType() != graph.DirectEdge && edge != lowestEdge {
+			if hasPathTo(edge, sourceType) {
```

Comment:
`if !hasPathTo(edge, sourceType) {return IntersectionEdges{}, nil}`

Simplifies the logic a bit more.